apiserver/facades/agent/keyupdater/authorisedkeys.go

```diff
@@ -52,100 +36,176 @@ var _ KeyUpdater = (*KeyUpdaterAPI)(nil)
 // functionality is added.
 func (api *KeyUpdaterAPI) WatchAuthorisedKeys(ctx context.Context, arg params.Entities) (params.NotifyWatchResults, error) {
 	results := make([]params.NotifyWatchResult, len(arg.Entities))
+	if len(arg.Entities) == 0 {
+		return params.NotifyWatchResults{Results: results}, nil
+	}
 
 	canRead, err := api.getCanRead()
 	if err != nil {
-		return params.NotifyWatchResults{}, err
+		return params.NotifyWatchResults{}, fmt.Errorf(
+			"checking can read status for key updater watch authorised keys: %w",
+			err,
+		)
 	}
+
 	for i, entity := range arg.Entities {
 		tag, err := names.ParseTag(entity.Tag)
 		if err != nil {
-			results[i].Error = apiservererrors.ServerError(err)
+			results[i].Error = apiservererrors.ParamsError(
+				params.CodeTagInvalid,
+				"cannot parse tag %q",
+				entity.Tag,
+			)
 			continue
 		}
-		// 1. Check permissions
+
+		if tag.Kind() != names.MachineTagKind {
+			results[i].Error = apiservererrors.ParamsError(
+				params.CodeTagKindNotSupported,
+				"tag %q unsupported, can only accept tags of kind %q",
+				tag, names.MachineTagKind,
+			)
+			continue
+		}
+
 		if !canRead(tag) {
-			results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
+			results[i].Error = apiservererrors.ParamsError(
+				params.CodeUnauthorized,
+				"does not have permission to read authorised keys for %q",
```

Comment:
Yeah I see what you are saying here. I have updated it. My preference would be to still keep the context of tag in the string. Not because the caller doesn't know who they are but because if we want to log this in low level RPC code or somewhere this context doesn't always exist we have it all there for the reader.

That is my thinking.

Reply:
No problem with tag in the string, was more about the "duplication" of the error cause phrase